packages/client/src/runtime/core/types/GetResult.ts

```diff
@@ -82,31 +81,25 @@ type SelectablePayloadFields<K extends PropertyKey, O> =
   | { composites: { [k in K]: O } }
 
 // prettier-ignore
-type SelectField<P extends SelectablePayloadFields<any, any>, K extends PropertyKey> = 
+type SelectField<P extends SelectablePayloadFields<any, any>, K extends PropertyKey> =
   P extends { objects: Record<K, any> } 
   ? P['objects'][K]
   : P extends { composites: Record<K, any> }
     ? P['composites'][K]
     : never
 
 // prettier-ignore
-export type DefaultSelection<P> = P extends Payload 
-  ? P['scalars'] & UnwrapPayload<P['composites']>
-  : P
+export type DefaultSelection<P> = UnwrapPayload<{ default: P }>['default']
 
 // prettier-ignore
-type UnwrapPayload<P> = {
-  [K in keyof P]:
-    P[K] extends Payload[]
-    ? UnwrapPayload<P[K]>
-    : P[K] extends Payload 
-      ? P[K]['scalars'] & UnwrapPayload<P[K]['composites']> 
-      : P[K] extends infer O | null 
-        ? O extends Payload 
-          ? (O['scalars'] & UnwrapPayload<O['composites']>) | null
-        : P[K]
-      : P[K]
-} & unknown
+export type UnwrapPayload<P> = {} extends P ? unknown : {
+  [K in keyof P]: 
+    P[K] extends { scalars: infer S, composites: infer C }[]
+    ? Array<S & UnwrapPayload<C>>
+    : P[K] extends { scalars: infer S, composites: infer C } | null
+      ? S & UnwrapPayload<C> | Select<P[K], null>
+      : never
+};
```

Comment:
This implementation should be more light weight as it does less comparisons similarly to `GetFindResult` does. It improved things significantly on the user-reproduction since immediate evaluation is guarded by `infer`.